### PR TITLE
Add conflict processor module

### PR DESCRIPTION
**Pull request by ajreynol**
This module can be used to minimize explanations of theory lemmas and conflicts. In practice, around 30% of lemmas can be minimized using this module.  This leads to significant improvements in performance in some industrial benchmarks.

A followup PR will optionally enable this technique (disabled by default, as it gives mixed performance on SMT-LIB).

Another use case will be for automatic proof reconstruction, as this module can infer proofs for theory lemmas in a generic way.

### CODE REVIEW

**Review by HanielB — CHANGES_REQUESTED** [review]

File: src/theory/conflict_processor.h [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  Node evaluateSubstitutionLit(const SubstitutionMap& s,
+                               const Node& tgtLit) const;
+  /**
+   * Is assignment equality? Returns true if n is an equality from which
+   * a substitution can be inferred and added to s. It does not consider
+   * substitutions that induce cycles or are for variables that already have
+   * substitutions. For example, given current substitution s = {y->z},
+   *    (= x y) returns true with x -> y.
+   *    (= x (f x)) return false.
+   *    (= y 0) returns false.
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/conflict_processor.h:118]: Yes, its a heuristic.

[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/conflict_processor.h:118]: FYI, I did start to write a more general method but quickly realized it was too heavy handed.  This is something to consider separately though.

File: src/theory/conflict_processor.h [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+   */
+  TrustNode processLemma(const TrustNode& lem);
+
+ private:
+  /** Common constants */
+  Node d_true;
+  Node d_false;
+  Node d_nullNode;
+  /** Use the extended rewriter? */
+  bool d_useExtRewriter;
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/conflict_processor.h:61]: Ah, somehow this was lost in merge, will add.

File: src/theory/conflict_processor.cpp [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+  visit.push_back(lem);
+  do
+  {
+    cur = visit.back();
+    visit.pop_back();
+    if (visited.find(cur) == visited.end())
+    {
+      visited.insert(cur);
+      k = cur.getKind();
+      if (k == Kind::OR || k == Kind::IMPLIES)
```
[the remark this answers is not in the document]
Comment (HanielB) [reply src/theory/conflict_processor.cpp:268]: Right, but won't this be only at the initial level, and not at every recursive case, which this loop is going through?

File: src/theory/conflict_processor.h [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  ConflictProcessor(Env& env);
+  ~ConflictProcessor() {}
+
+  /**
+   * Attempt to rewrite a lemma to a stronger one. For example, the lemma
+   * (=> (= x a) (or B C)) may be replaced by (=> (= x a) B) if B[a/x] rewrites
+   * to true.
+   * 
+   * @param lem The lemma.
+   * @return A trust node for a lemma that implies lem.
```
[the remark this answers is not in the document]
Comment (HanielB) [reply src/theory/conflict_processor.h:58]: ping

File: src/theory/conflict_processor.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/conflict_processor.cpp:268]: you could also have `(or (or A B (or C D) E) F)` in theory.

**Review by HanielB — APPROVED** [review]

**Pull request merged**